packages/sync-actions/src/utils/copy-empty-array-props.js

```diff
@@ -13,46 +13,41 @@ const CUSTOM = 'custom'
 export default function copyEmptyArrayProps(oldObj = {}, newObj = {}) {
   if (!isNil(oldObj) && !isNil(newObj)) {
     const nextObjectWithEmptyArray = Object.entries(oldObj).reduce(
-      (nextObject, [key, value]) => {
-        const merged = {
-          ...newObj,
-          ...nextObject,
-        }
-
+      (merged, [key, value]) => {
         // Ignore CUSTOM key as this object is dynamic and its up to the user to dynamically change it
         // todo, it would be better if we pass it as ignored keys param
         if (key === CUSTOM) return merged
 
         if (Array.isArray(value) && newObj[key] && newObj[key].length >= 1) {
           /* eslint-disable no-plusplus */
+          const hashMapValue = value.reduce((acc, val) => {
+            acc[val.id] = val
+            return acc
+          }, {})
```

Comment:
This is a small trick to avoid the usage of `find` in a situation of array of array (of arrays). The difference between both approaches was remarkable when debugging this locally.

With hashmap it was less than 100 ms
With `.find` around 6 seconds

Reply:
As a pattern I will name this trick a "dictionary"... you could even do this a little trickier

```
 const valueDict = Object.assign({}, ...value.map( x=>( { [x.id]: x) ) )
 ```

Reply:
fwiw the object assign method is a bit slower, probably because you are mapping through all of the values three times (once for the map, once for the spread, once for the assignment).

If it's a large array (100k elements) the reduce approach takes ~2ms whereas the object assign approach takes ~50ms.

```
// Create an array for testing.
const testArray = new Array(100_000).fill(null).map((_, index) => ({ id: index }));

// Store individual iteration times.
const reduceTimes = [];
const objectAssignTimes = [];

const testIterations = 1_000;
for (let i = 0; i < testIterations; i++) {
  // Run reduce method.
  const startTimeReduce = Date.now();
  testArray.reduce((acc, element) => {
    acc[element.id] = element;
    return acc;
  }, {});
  reduceTimes.push(Date.now() - startTimeReduce);

  // Run object assign method.
  const startTimeObjectAssign = Date.now();
  Object.assign({}, ...testArray.map((element) => ({ [element.id]: element })));
  objectAssignTimes.push(Date.now() - startTimeObjectAssign);
}

// Log results.
console.log({ totalTimeReduce: reduceTimes.reduce((acc, val) => acc + val) / testIterations });
console.log({
  totalTimeObjectAssign: objectAssignTimes.reduce((acc, val) => acc + val) / testIterations,
});
```

Reply:
Well... I just said trickier :-)

Thanks for the testing!

Another interesting approach will be if Array.prototype.group proposal gets approved. It won't return exactly a dictionary because the values will be an array...

But it will read nicely:

```
array.group( x=>x.id ) 
```

Reply:
Thanks both for the input and the testing around this. It was really a nice learning experience and I think this improves the library for those cases that we see big numbers.

Just for the record, the case that we were checking was a customer from cimpress that had almost 13k addresses defined in it and the size of the entity was almost 9MB 😨 💥